walk/walk.go

```diff
@@ -162,32 +162,38 @@ func Walk(c *config.Config, cexts []config.Configurer, dirs []string, mode Mode,
 			}
 		}
 
-		shouldUpdate := shouldUpdate(rel, mode, updateParent, updateRels)
+		shouldUpdate := updateRels.shouldUpdate(rel, updateParent)
 		for _, sub := range subdirs {
-			if subRel := path.Join(rel, sub); shouldVisit(subRel, mode, shouldUpdate, updateRels) {
+			if subRel := path.Join(rel, sub); updateRels.shouldVisit(subRel, shouldUpdate) {
 				visit(c, filepath.Join(dir, sub), subRel, shouldUpdate)
 			}
 		}
 
 		update := !haveError && !wc.ignore && shouldUpdate
-		if shouldCall(rel, mode, updateParent, updateRels) {
+		if updateRels.shouldCall(rel, updateParent) {
 			genFiles := findGenFiles(wc, f)
 			wf(dir, rel, c, update, f, subdirs, regularFiles, genFiles)
 		}
 	}
 	visit(c, c.RepoRoot, "", false)
 }
 
-// buildUpdateRelMap builds a table of prefixes, used to determine which
+// An updateFilter tracks which directories need to be updated
+type updateFilter struct {
+	mode Mode
+
+	// map from slash-separated paths relative to the
+	// root directory ("" for the root itself) to a boolean indicating whether
+	// the directory should be updated.
+	updateRels map[string]bool
+}
+
+// NewUpdateFilter builds a table of prefixes, used to determine which
```

Comment:
Could you mark this as internal via a comment?

Reply:
I've added an INTERNAL note.